File: Algorithm/C++/1000+/1915. Number of Wonderful Substrings.cc
#include <string>
#include <unordered_map>

class Solution {
public:
    long long wonderfulSubstrings(std::string word) {
        long long result = 0;
        int n = word.size(), prefix = 0, tmp;
        std::unordered_map<int, int> counter;
        counter[0] = 1;
        for (char c: word) {
            prefix = prefix xor (1 << (c-'a'));
            if (counter.find(prefix) != counter.end()) {
                result += counter[prefix];
                counter[prefix]++;
            } else {
                counter[prefix]=1;
            }
            for (int i=0; i < 10; i++) {
                tmp = prefix xor (1 << i);
                if (counter.find(tmp) != counter.end()) {
                    result += counter[tmp];
                }
            }
        }
        return result;
    }
};

File: Algorithm/Python/1000+/1915. Number of Wonderful Substrings.py
# prefix sum and bit manipulation, time complexity: O(n)
from collections import defaultdict

class Solution(object):
    def wonderfulSubstrings(self, word):
        """
        :type word: str
        :rtype: int
        """
        n, result = len(word), 0
        counter, prefix = defaultdict(int), 0
        counter[0] = 1
        for i in range(n):
            prefix ^= (1 << (ord(word[i]) - ord('a')))
            result += counter[prefix]
            counter[prefix] += 1
            for j in range(10):
                tmp = prefix^(1<<j)
                if tmp in counter:
                    result += counter[tmp]
        return result

#################################################
# naive solution, time complexity:O(n^2), TLE
from collections import defaultdict

class Solution(object):
    def wonderfulSubstrings(self, word):
        """
        :type word: str
        :rtype: int
        """
        n, result = len(word), 0
        for i in range(n):
            counter, odd_cnt = defaultdict(int), 0
            for j in range(i, n):
                counter[word[j]] += 1
                if (counter[word[j]] & 1) == 1:
                    odd_cnt += 1
                elif counter[word[j]] > 1:
                    odd_cnt -= 1
                if odd_cnt <= 1:
                    result += 1
        return result
                


